internal/statemachine/helper.go

```diff
@@ -354,3 +371,150 @@ func maxOffset(offset1, offset2 quantity.Offset) quantity.Offset {
+				Bootable: bootable,
+			}
+			mbrPartitions = append(mbrPartitions, mbrPartition)
+		} else {
+			partitionType := gpt.Type(structureType)
+			gptPartition := &gpt.Partition{
+				Start: uint64(*structure.Offset) / sectorSize,
+				Size:  uint64(structure.Size),
+				Type:  partitionType,
+				Name:  structure.Name,
+			}
+			gptPartitions = append(gptPartitions, gptPartition)
+		}
+	}
+
+	if volume.Schema == "mbr" {
+		mbrTable := &mbr.Table{
+			Partitions:         mbrPartitions,
+			LogicalSectorSize:  int(sectorSize),
+			PhysicalSectorSize: int(sectorSize),
+		}
+		partitionTable = mbrTable
+	} else {
+		gptTable := &gpt.Table{
+			Partitions:         gptPartitions,
+			LogicalSectorSize:  int(sectorSize),
+			PhysicalSectorSize: int(sectorSize),
+			ProtectiveMBR:      true,
+		}
+		partitionTable = gptTable
+	}
+	return &partitionTable
+}
+
+// calculateImageSize calculates the total sum of all partition sizes in an image
+func (stateMachine *StateMachine) calculateImageSize() (int64, error) {
+	if stateMachine.GadgetInfo == nil {
+		return 0, fmt.Errorf("Cannot calculate image size before initializing GadgetInfo")
+	}
+	var imgSize int64 = 0
+	for _, volume := range stateMachine.GadgetInfo.Volumes {
+		for _, structure := range volume.Structure {
+			imgSize += int64(structure.Size)
+		}
+	}
+	return imgSize, nil
+}
+
+// copyDataToImage runs dd commands to copy the raw data to the final image with appropriate offsets
+func (stateMachine *StateMachine) copyDataToImage(volumeName string, volume *gadget.Volume, diskImg *disk.Disk) error {
+	for structureNumber, structure := range volume.Structure {
+		if shouldSkipStructure(structure, stateMachine.isSeeded) {
+			continue
+		}
+		sectorSize := diskImg.LogicalBlocksize
+		// set up the arguments to dd the structures into an image
+		partImg := filepath.Join(stateMachine.tempDirs.volumes, volumeName,
+			"part"+strconv.Itoa(structureNumber)+".img")
+		seek := strconv.FormatInt(int64(getStructureOffset(structure))/sectorSize, 10)
+		count := strconv.FormatFloat(math.Ceil(float64(structure.Size)/float64(sectorSize)), 'f', 0, 64)
+		ddArgs := []string{
+			"if=" + partImg,
+			"of=" + diskImg.File.Name(),
+			"bs=" + strconv.FormatInt(sectorSize, 10),
+			"seek=" + seek,
+			"count=" + count,
+			"conv=notrunc",
+			"conv=sparse",
+		}
+		if err := helperCopyBlob(ddArgs); err != nil {
+			return fmt.Errorf("Error writing disk image: %s",
+				err.Error())
+		}
+	}
+	return nil
+}
+
+// writeOffsetValues handles any OffsetWrite values present in the volume structures.
+func writeOffsetValues(volume *gadget.Volume, imgName string, sectorSize, imgSize uint64) error {
+	imgFile, err := osOpenFile(imgName, os.O_RDWR, 0755)
+	if err != nil {
+		return fmt.Errorf("Error opening image file to write offsets: %s", err.Error())
+	}
+	defer imgFile.Close()
+	for _, structure := range volume.Structure {
+		if structure.OffsetWrite != nil {
+			offset := uint64(*structure.Offset) / sectorSize
```

Comment:
...while THIS, on the other hand, IS correct! huh, seems like we didn't quite have consistency in handling offsets and sizes when counted as sectors! Weird!